RTIO: Multishot: Stop resubmitting multishot items on error

ubieda opened the pull request:
### Description

This PR adds error-handling to Multi-shot items by not resubmitting them on error occurrence.

This PR includes:
- Refactoring prior to including functionality, in order to improve readability.
- Aborting Multi-shot items upon its occurrence.
- Testcase locking in behavior.

bjarki-andreasen commented:
> > I wonder if it may be simpler for the user to check the result of and manually resubmit the SQE, given I believe an SQE is cancelled if there is an error.
> > `RTIO_SQE_MULTISHOT_BLOCKED` effectively dequeues the SQE until the user allows it to be re-enqueued, is this not essentially the same as the SQE being cancelled on error (dequeued), and the user resubmitting the SQE later?
> 
> AFAIK, cancelled items cannot be _un-cancelled_ (as in, re-use the submission), so the client needs to re-create the SQE submission from the top.
> 
> However, I see a couple pros of having a separate flag instead of reusing the existing one:
> 
>     * I see value in keeping the authority of cancelling items to the clients only mainly for troubleshooting purposes.
> 
>     * I also like that the flag cannot be confused by other one-shot items that have been cancelled (e.g: in a scenario of a single RTIO client working with multiple IODEVs).
> 
> 
> At the end of the day, what we want to do is to let the user know a multi-shot failed, and it's on them to handle it in order to recover. I'm game for discussing what's the best way to do it if it's not this one.

I thought chained SQEs where automatically cancelled if any SQE fails, I would imagine the same behavior for a multishot SQE (basically an infinite SQE chain), maybe this is not the case?

teburd commented:
Multi shot is a bit odd here but I’m with Bjarki, if a multi shot fails a failed completion should be the result and it should not be automatically resubmitted is my thinking.

ubieda committed:
rtio: Assert multishot flag is not combined with txn or chained

There's no use-case covering this case and enforcing it will allow
simplify the logic for processing items with each of these items when
mutually exclusive.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
rtio: executor: split handling of one-shot and multi-shot completions

As their handling is not that similar and mixing them both affects
code readability. This patch does not affect functionality and does
not break any testsuite under rtio API.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
rtio: multishot: Stop resubmitting multishot items on error

As an error-handling strategy. The client then can recover from this
depending on the error and the number of occurrences.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: rtio: iodev: fix: Add missing early return when error occurs

It appears this had not been exposed as the testsuite does not
exercise forcing errors for the IODEV. However, in the upcoming
testcase for error-handling of multi-shot items, it will be required.

Placed as standalone patch for the sake of clarity.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: rtio: Add testcase to demonstrate multishot aborted on errors

By forcing IODEV test to throw errors and verifying no more CQE's are
coming through after the forced error.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

bjarki-andreasen reviewed (dismissed):
Looks good!

teburd commented on subsys/rtio/rtio_executor.c:
```diff
@@ -122,71 +125,108 @@ void rtio_executor_submit(struct rtio *r)
+		 * the corresponding error: whether re-submitting,
+		 * rebooting or anything else.
+		 */
+		iodev_sqe->sqe.flags |= RTIO_SQE_CANCELED;
```
Why do we mark an errored sqe as cancelled?

teburd replied:
I'm a bit lost on this. This branch doesn't seem to be needed unless I've missed something.

All this flag is going to do is ensure the if (is_cancelled) branch below takes place, which reads wrong to me. The branch below should be reading...

If the submission has been cancelled *or* has failed, we free the resources. We should be doing that *before* notifying the completion.

Imagine I do

```C
RTIO_DEFINE(r, 1, 1)
```

submit a multishot, get the completion (errored), and now try and submit here, I won't have the submission available to me.

Is there some other place I'm not seeing that this flag is looked at for multishot submissions?

ubieda replied:
Resolved to just release resources on error, without cancelling (as suggested) If we need to come back to this we can revisit.

teburd replied:
Ah you are thinking of the handles case. With user mode the thread won’t see updates to the sqe as it’s a copy. Only the user data connects the sqe and cqe together, really cancelled is a bit odd and changed a lot of things to make the sqe mutable instead of immutable as they were originally

teburd commented on subsys/rtio/rtio_executor.c:
```diff
@@ -122,71 +125,108 @@ void rtio_executor_submit(struct rtio *r)
 	}
-	if (!is_canceled) {
+
+	if (is_canceled) {
```
is_cancelled || !is_ok and then you don’t need to mutate the sqe

teburd replied:
This branch should also likely occur before the cqe_submit

ubieda replied:
Addressed.

ubieda commented:
Sorry - I didn't rebased correctly. Hold on a sec
